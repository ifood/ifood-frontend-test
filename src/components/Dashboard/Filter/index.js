/* eslint-disable react/jsx-props-no-spreading */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { FaSearch, FaFilter, FaAngleDown, FaAngleUp } from 'react-icons/fa';
import Select from 'react-select';
import DatePicker, { registerLocale } from 'react-datepicker';
import pt from 'date-fns/locale/pt';

import useSkipFirstRender from '../../../utils/useSkipFirstRender';

import { Container, Label } from './styles';
import 'react-datepicker/dist/react-datepicker.css';

registerLocale('pt', pt);

function Filter({
  playlists,
  setPlaylists,
  setFilters,
  getPlaylistsService,
  filtersLabelValue,
}) {
  const [localeOptions, setLocaleOptions] = useState([]);
  const [countryOptions, setCountryOptions] = useState([]);
  const [playlistsInitial, setPlaylistsInitial] = useState([]);
  const [values, setValues] = useState({});
  const [isShowingFilter, setIsShowingFilter] = useState(false);
  const [searchValueFocus, setSearchValueFocus] = useState(false);

  const handleSearchInputChange = (event) => {
    event.persist();

    const searchValue = event.target.value;

    if (searchValue.length) {
      const filteredItems = playlistsInitial.filter(({ name }) =>
        name.toLocaleLowerCase().includes(searchValue.toLocaleLowerCase()),
      );

      setValues((prevState) => ({ ...prevState, searchValue }));
      setPlaylists(filteredItems);
    } else {
      setPlaylists(playlistsInitial);
      setValues((prevState) => ({ ...prevState, searchValue: '' }));
    }
  };

  const handleInputChange = (name, value) => {
    setValues((prevState) => ({ ...prevState, [name]: value }));
  };

  const handleDateTimeChange = (value) => {
    setValues((prevState) => ({ ...prevState, date: value }));
  };

  useSkipFirstRender(() => {
    const { searchValue, locale, country, date, limit, offset } = values;

    setFilters({
      searchValue,
      locale: locale || 'pt_BR',
      country: country || 'BR',
      timestamp: (date && date.toISOString()) || new Date().toISOString(),
      limit: limit || 12,
      offset: offset || 0,
    });
    getPlaylistsService();
  }, [values]);

  useEffect(() => {
    if (!playlistsInitial.length) {
      setPlaylistsInitial(playlists);
    }

    if (filtersLabelValue) {
      let locales = filtersLabelValue.find((value) => value.id === 'locale');
      let countries = filtersLabelValue.find((value) => value.id === 'country');

      locales = locales.values.map((item) => ({ ...item, label: item.name }));
      countries = countries.values.map((item) => {
        let { value } = item;
        let label = item.name;

        if (value === 'en_US') {
          value = 'US';
          label = 'Estados Unidos';
        }

        return { value, label };
      });

      setLocaleOptions(locales);
      setCountryOptions(countries);
    }
  }, [playlists, playlistsInitial, filtersLabelValue]);

  return (
    <Container showFilter={isShowingFilter}>
      <div className="filter-wrapper">
        <div className="filter-content">
          <div className="filter-search-container">
            <Label focus={searchValueFocus}>
              <FaSearch size={16} className="fa fa-search" />
              <input
                name="searchValue"
                type="text"
                placeholder="Filtrar"
                onChange={handleSearchInputChange}
                onFocus={() => setSearchValueFocus(true)}
                onBlur={() => setSearchValueFocus(false)}
              />
            </Label>
            <button
              type="button"
              className="filter-button"
              onClick={() => setIsShowingFilter(!isShowingFilter)}
            >
              <FaFilter size={16} className="fa fa-filter" />
              {isShowingFilter ? (
                <FaAngleUp size={18} className="fa fa-angle-down" />
              ) : (
                <FaAngleDown size={18} className="fa fa-angle-down" />
              )}
            </button>
          </div>
          <div className="filter-items">
            <Select
              name="locale"
              className="react-select"
              classNamePrefix="select"
              placeholder="Localidade"
              isSearchable={false}
              options={localeOptions}
              onChange={(value) => handleInputChange('locale', value.value)}
            />
            <Select
              name="country"
              className="react-select"
              classNamePrefix="select"
              placeholder="País"
              isSearchable={false}
              options={countryOptions}
              onChange={(value) => handleInputChange('country', value.value)}
            />
            <DatePicker
              showTimeInput
              locale="pt"
              timeInputLabel="Horário:"
              timeFormat="p"
              dateFormat="Pp"
              placeholderText="Dia e horário"
              className="default-input"
              selected={values.date}
              onChange={(date) => handleDateTimeChange(date)}
            />
            <input
              name="limit"
              type="number"
              placeholder="Itens por página"
              className="default-input"
              onChange={(e) =>
                e.target.value <= 60 &&
                handleInputChange(e.target.name, e.target.value)
              }
              value={values.limit}
            />
            <input
              name="offset"
              type="number"
              placeholder="Página"
              className="default-input"
              onChange={(e) => handleInputChange(e.target.name, e.target.value)}
              value={values.offset}
            />
          </div>
        </div>
      </div>
    </Container>
  );
}

Filter.defaultProps = {
  filtersLabelValue: [],
};

Filter.propTypes = {
  filtersLabelValue: PropTypes.arrayOf(PropTypes.any),
  playlists: PropTypes.arrayOf(PropTypes.any).isRequired,
  setPlaylists: PropTypes.func.isRequired,
  setFilters: PropTypes.func.isRequired,
  getPlaylistsService: PropTypes.func.isRequired,
};

export default Filter;
