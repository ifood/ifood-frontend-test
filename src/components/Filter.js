// import { useEffect, useState } from 'react'
import usePlaylists from '../contexts/PlaylistContext'
import FieldsFilter from './FieldsFilter'

const Filter = () => {
  const { filterFields, loadingFilterFields } = usePlaylists()

  return (
    <>
      {loadingFilterFields ? (
        <p>Loading Filter</p>
      ) : (
        <div>
          {filterFields.map((filter) => (
            <div key={filter.id}>
              <h3>{filter.name}</h3>
              <FieldsFilter filter={filter} />
              {/* <ul>
                {filter.values &&
                  filter.values.map((option) => <p>{option.name}</p>)}
              </ul> */}
            </div>
          ))}
        </div>
      )}
    </>
  )
}

export default Filter
