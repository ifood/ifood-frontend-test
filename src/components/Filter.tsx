import React, { PureComponent } from 'react';

interface IProps {
  onSignOut: () => void;
}

export default class Filter extends PureComponent<IProps> {
  public render() {
    return (
      <div>
        <h1>Hi, I'm the filter component</h1>
        <input type='text' />
        <select />
        <button onClick={this.props.onSignOut}>Sign out</button>
      </div>
    );
  }
}
