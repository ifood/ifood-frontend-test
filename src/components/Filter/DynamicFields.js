import React from 'react';
import FieldContainer from '../shared/FieldContainer';
import Select from '../shared/Select';
import Input from '../shared/Input';

const DynamicFields = ({
  fields = [],
  onChange,
  formData = {}
 }) => {
  if (!fields.length) return null;
  return fields.map(field => {
    const filterItem = field.values ?
      <Select
        values={field.values}
        name={field.id}
        onChange={onChange}
        selectedValue={formData[field.id]}
      /> :
      <Input
        validation={field.validation}
        type={field.validation.primitiveType}
        name={field.id}
        onChange={onChange}
        value={formData[field.id]}
        placeholder={field.validation.pattern || field.name}
      />;
    return (
      <FieldContainer name={field.name} key={field.id}>
        { filterItem }
      </FieldContainer>
    )
  });
}

export default DynamicFields;