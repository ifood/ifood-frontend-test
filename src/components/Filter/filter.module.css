
.title {
  margin: -15px -15px 15px;
  background-color: #007bff;
  padding: 15px 10px;
  color: #ffffff;
  font-size: 1.5rem;
}

.form {
  padding: 15px;
  padding-bottom: 0;
}