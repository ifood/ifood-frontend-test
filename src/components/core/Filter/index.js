import React from "react";
import PropTypes from "prop-types";
import { useForm, FormContext } from "react-hook-form";

import { handleInputMask } from "helpers/masks";

import { CustomColumns } from "components/core/Grid";
import Button from "components/core/Button";
import Select from "components/core/Select";
import Loading from "components/core/Loading";

import * as S from "./styles";

const Filter = ({ isLoading, options, handleFilter }) => {
  const { register, handleSubmit, ...methods } = useForm({
    mode: "onChange",
    reValidateMode: "onBlur",
    submitFocusError: true,
  });

  const { setValue } = methods;

  return (
    <Loading isLoading={isLoading}>
      <S.Div>
        <FormContext setValue={setValue} {...methods}>
          <form autoComplete="off" onSubmit={handleSubmit(handleFilter)}>
            <CustomColumns minmax="150px">
              {options.map((option) => (
                <S.Wrapper key={option.id}>
                  {!option.values ? (
                    <S.Field
                      name={option.id}
                      id={option.id}
                      type="text"
                      ref={register}
                      onInput={(e) => {
                        option?.validation?.primitiveType &&
                          handleInputMask(
                            e,
                            option.validation.primitiveType,
                            option.validation?.entityType
                          );
                      }}
                    />
                  ) : (
                    <Select
                      register={register}
                      optionId={option.id}
                      values={option.values}
                    />
                  )}
                  <S.Label htmlFor={option.id}>{option.name}</S.Label>

                  {option.validation?.pattern && (
                    <S.Format>Formato: {option.validation?.pattern}</S.Format>
                  )}
                </S.Wrapper>
              ))}
            </CustomColumns>

            <Button margin="1rem 0 0">Filtrar</Button>
          </form>
        </FormContext>
      </S.Div>
    </Loading>
  );
};

Filter.propTypes = {
  isLoading: PropTypes.bool,
  handleFilter: PropTypes.func.isRequired,
  options: PropTypes.array,
};

Filter.defaultProps = {
  isLoading: false,
  options: [],
};

export default Filter;
