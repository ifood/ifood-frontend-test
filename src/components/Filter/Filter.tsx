import React, { useEffect, useState } from 'react';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import Collapse from '@material-ui/core/Collapse';
import Grid from '@material-ui/core/Grid';
import InputAdornment from '@material-ui/core/InputAdornment';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import FilterListIcon from '@material-ui/icons/FilterList';
import SearchIcon from '@material-ui/icons/Search';
import { FilterField as FilterFieldComponent } from './FilterField/FilterField';
import { useFilterFields, FilterField } from '../../hooks/useFilterFields';
import { useStyles } from './Filter.jss';

const getFilterFieldInitialValue = (filterField: FilterField) => {
  return filterField.type === 'datetime' ? null : '';
};

interface FilterItem {
  filterField: FilterField;
  value: string | null;
  error?: string;
}

interface FilterProps {
  onFilter: (filters: Record<string, string>) => void;
  onSearch: (search: string) => void;
}

export const Filter: React.FC<FilterProps> = ({ onFilter, onSearch }) => {
  const classes = useStyles();
  const [filters, setFilters] = useState<FilterItem[]>([]);
  const [isExpanded, setIsExpanded] = useState(false);
  const { filterFields, isLoading, isSuccess, isError, validateFilterField } = useFilterFields();

  const hasError = filters.some((filter) => filter.error);

  useEffect(() => {
    if (filterFields) {
      setFilters(
        filterFields.map((filterField) => ({
          filterField,
          value: getFilterFieldInitialValue(filterField),
          error: undefined,
        }))
      );
    } else {
      setFilters([]);
    }
  }, [filterFields]);

  const clearFilters = () => {
    setFilters((state) =>
      state.map((filter) => ({
        ...filter,
        value: getFilterFieldInitialValue(filter.filterField),
        error: undefined,
      }))
    );
  };

  const toggleExpanded = () => {
    setIsExpanded(!isExpanded);
  };

  const handleChangeSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
    onSearch(event.target.value);
  };

  const handleChangeFilter = (id: string, value: string | null) => {
    setFilters((state) =>
      state.map((filter) => {
        if (filter.filterField.id === id) {
          return {
            ...filter,
            value,
            error: validateFilterField(id, value),
          };
        }
        return { ...filter };
      })
    );
  };

  const handleClearFilters = () => {
    clearFilters();
    onFilter({});
  };

  const handleApplyFilters = () => {
    onFilter(
      filters.reduce((filterValues: Record<string, string>, filter) => {
        if (filter.value) {
          filterValues[filter.filterField.id] = filter.value;
        }
        return filterValues;
      }, {})
    );
  };

  return (
    <Paper className={classes.container}>
      <div className={classes.search}>
        <TextField
          variant="outlined"
          placeholder="Buscar playlists..."
          size="small"
          onChange={handleChangeSearch}
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <SearchIcon color="action" />
              </InputAdornment>
            ),
          }}
        />
        <Button color="primary" startIcon={<FilterListIcon />} onClick={toggleExpanded}>
          Filtros
        </Button>
      </div>

      <Collapse in={isExpanded} timeout="auto">
        {isLoading && (
          <div className={classes.indicator}>
            <CircularProgress color="primary" />
          </div>
        )}

        {isError && (
          <div className={classes.indicator}>
            <Typography>Não foi possível carregar os filtros.</Typography>
          </div>
        )}

        {isSuccess && filters.length === 0 && (
          <div className={classes.indicator}>
            <Typography>Nenhum filtro encontrado.</Typography>
          </div>
        )}

        {isSuccess && filters.length > 0 && (
          <Grid className={classes.filters} container spacing={2}>
            {filters.map((filter) => (
              <Grid item xs={12} sm={4} md={3} key={filter.filterField.id}>
                <FilterFieldComponent
                  filterField={filter.filterField}
                  value={filter.value}
                  error={filter.error}
                  onChange={(value) => handleChangeFilter(filter.filterField.id, value)}
                />
              </Grid>
            ))}
            <Grid item xs={12} sm={4} md={3}>
              <div className={classes.actions}>
                <Button color="primary" variant="outlined" onClick={handleClearFilters}>
                  Limpar
                </Button>
                <Button
                  color="primary"
                  variant="contained"
                  disableElevation
                  disabled={hasError}
                  onClick={handleApplyFilters}
                >
                  Filtrar
                </Button>
              </div>
            </Grid>
          </Grid>
        )}
      </Collapse>
    </Paper>
  );
};
