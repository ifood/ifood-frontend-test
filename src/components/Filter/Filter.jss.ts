import { makeStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles(
  (theme: Theme) => ({
    container: {
      margin: theme.spacing(2, 2, 0),
      padding: theme.spacing(2),
    },
    search: {
      display: 'flex',
      alignItems: 'center',
      '& > :first-child': {
        flexGrow: 1,
        marginRight: theme.spacing(2),
      },
    },
    filters: {
      marginTop: theme.spacing(1),
    },
    actions: {
      display: 'flex',
      alignItems: 'center',
      height: '40px',
      '& > *': {
        flexBasis: 0,
        flexGrow: 1,
        height: '100%',
      },
      '& > :not(:last-child)': {
        marginRight: theme.spacing(1),
      },
    },
    indicator: {
      marginTop: theme.spacing(2),
      display: 'flex',
      justifyContent: 'center',
    },
  }),
  { name: 'Filter' }
);
