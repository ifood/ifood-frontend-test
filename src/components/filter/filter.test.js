import React from 'react';
import Filter from './filter';
import { shallow } from 'enzyme';

// Mounting - Full DOM rendering including child components
// Shallow - Renders only the single component, not including its children,
// this is useful to isolate the component for pure unit testing.
// Render - Renders to static HTML, including children

it('should render correctly with no props', () => {
  const component = shallow(<Filter/>);
  expect(component).toMatchSnapshot();
});

it('should render correctly with props null', () => {
    const component = shallow(<Filter onTextChange={null} onChangeFilters={null} filterValues={null}/>);
    expect(component).toMatchSnapshot();
});

it('should render correctly with props', () => {
  const filters = {
    locale: 'sv_SE',
    country: 'SE',
    datetime: '2014-10-23T09:00',
    limit: '1',
    offset: '3'
  }

  const component = shallow(<Filter onTextChange={null} onChangeFilters={null} filterValues={filters}/>);
  expect(component).toMatchSnapshot();
});

