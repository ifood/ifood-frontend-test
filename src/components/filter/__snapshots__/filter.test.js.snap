// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with no props 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextField
      id="search"
      label="Filter playlists by name"
      onKeyUp={[Function]}
      required={false}
      select={false}
      style={
        Object {
          "marginBottom": ".8em",
        }
      }
      variant="standard"
    />
  </div>
</div>
`;

exports[`should render correctly with props 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextField
      id="search"
      label="Filter playlists by name"
      onKeyUp={[Function]}
      required={false}
      select={false}
      style={
        Object {
          "marginBottom": ".8em",
        }
      }
      variant="standard"
    />
  </div>
</div>
`;

exports[`should render correctly with props null 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextField
      id="search"
      label="Filter playlists by name"
      onKeyUp={[Function]}
      required={false}
      select={false}
      style={
        Object {
          "marginBottom": ".8em",
        }
      }
      variant="standard"
    />
  </div>
</div>
`;
